File: bindings/dart/bao/lib/src/mailbox.dart
import 'package:bao/src/bindings.dart';
import 'package:bao/src/message.dart';

class Mailbox {
  int vaultHandle;

  Mailbox(this.vaultHandle);

  Future<void> send(String dir, String group, Message message) async {
    var res = await bindings.acall(
        'bao_mailbox_send', [vaultHandle, dir, group, message]);
    res.throwIfError();
  }

  Future<List<Message>> receive(String dir,
      {DateTime? since, int fromId = 0}) async {
    var res = await bindings.acall('bao_mailbox_receive',
        [vaultHandle, dir, since?.millisecondsSinceEpoch ?? 0, fromId]);
    return res.list.map((x) => Message.fromJson(x)).toList();
  }

  Future<void> download(
      String dir, Message message, int attachmentIndex, String dest) async {
    var res = await bindings.acall(
        'bao_mailbox_download', [vaultHandle, dir, message, attachmentIndex, dest]);
    res.throwIfError();
  }
}

File: bindings/java/lib/src/main/java/ink/francesco/bao/Mailbox.java
package ink.francesco.bao;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.List;
import java.util.Map;

public class Mailbox {

    public static class Message {
        public String subject;
        public String body;
        public List<String> attachments;
        public Map<String, Object> fileInfo;
    }

    private final Vault bao;
    private final ObjectMapper mapper = new ObjectMapper();

    public Mailbox(Vault bao) {
        this.bao = bao;
    }

    public void send(String dir, String group, Message message) throws Exception {
        Result r = BaoLibrary.instance.bao_mailbox_send(bao.hnd, dir, group, mapper.writeValueAsString(message));
        r.check();
    }

    public List<Message> receive(String dir, long since, long fromId) throws Exception {
        Result r = BaoLibrary.instance.bao_mailbox_receive(bao.hnd, dir, since, fromId);
        return r.list(Message.class);
    }

    public void download(String dir, Message message, int attachmentIndex, String dest) throws Exception {
        Result r = BaoLibrary.instance.bao_mailbox_download(bao.hnd, dir, mapper.writeValueAsString(message), attachmentIndex, dest);
        r.check();
    }
}
